api/pkg/transformer/feast/udf.go

```diff
@@ -0,0 +1,110 @@
+package feast
+
+import (
+	"errors"
+	"fmt"
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/vm"
+	"github.com/mmcloughlin/geohash"
+	"github.com/oliveagle/jsonpath"
+	"reflect"
+	"strconv"
+)
+
+type UdfResult struct {
+	Value interface{}
+	Error error
+}
+
+type UdfEnv struct {
+	UnmarshalledJsonRequest interface{}
+}
+
+func (env UdfEnv) Geohash(latitudeJsonPath, longitudeJsonPath string) UdfResult {
+	value, err := extractGeohash(env.UnmarshalledJsonRequest, latitudeJsonPath, longitudeJsonPath)
+	return UdfResult{
+		Value: value,
+		Error: err,
+	}
+}
+
+func toList(o interface{}) []interface{} {
+	var listInterface []interface{}
+	switch o.(type) {
+	case []interface{}:
+		listInterface = o.([]interface{})
+	case interface{}:
+		listInterface = []interface{}{o}
+	}
+	return listInterface
+}
+
+func toFloat64(o interface{}) (float64, error) {
+	var floatValue float64
+	switch o.(type) {
+	case float64:
+		floatValue = o.(float64)
+	case string:
+		parsed, err := strconv.ParseFloat(o.(string), 64)
+		if err != nil {
+			return 0, err
+		}
+		floatValue = parsed
+	case int:
+		floatValue = float64(o.(int))
+	default:
+		return 0, errors.New(fmt.Sprintf("%v cannot be parsed into float64", reflect.TypeOf(o)))
+	}
+	return floatValue, nil
+
+}
+
+func extractGeohash(jsonRequestBody interface{}, latitudeJsonPath, longitudeJsonPath string) (interface{}, error) {
+	latitude, err := jsonpath.JsonPathLookup(jsonRequestBody, latitudeJsonPath)
```

Comment:
`jsonpath.JsonPathLookup` will call json path Compile, it will be better if the compiled json path can be cached.

Reply:
Yes, it's definitely better to cache compiled json path. But implementation wise, that's a little tricky: the json path for latitude and longitude are part of the UDF expression. If we want to stored the compiled json path during transformer initialization, we would have to somehow extract the arguments from the expression, then store it in a map. It is also hard to generalize to all udfs, and the udf would have different arguments.

Reply:
Do we know how much latency is introduced by compiling the json path on the fly? If it is very significant, i will try to figure out some ways to cache the json path, but probably as a separate pull request?

Reply:
I am not too sure the additional processing time, but you can lazily create cache in this method itself, thus, only the first call to this method will call `Compile`. You'll need to ensure it's concurrency-safe.